### PR TITLE
Runtime TP

### PR DESCRIPTION
**Pull request by grimoire**
Support Both deploy-time TP and runtime TP
Support GQA runtime TP
Code is so ugly.

### CODE REVIEW

File: lmdeploy/turbomind/turbomind.py [hunk 1]
```diff
@@ -71,20 +71,22 @@ class TurboMind:
...
         eos_id (int): eos token id
+        tp (int): tensor parallel
     """
 
     def __init__(self,
                  model_path: str,
                  data_type: str = 'fp16',
-                 eos_id: int = 2):
+                 eos_id: int = 2,
+                 tp: int = torch.cuda.device_count()):
```
Comment (tpoisonooo) [inline lmdeploy/turbomind/turbomind.py:81]: 我倾向于默认值是 1，因为用 100B 概率小于用 7B 的。

File: lmdeploy/turbomind/turbomind.py [hunk 2]
```diff
@@ -97,9 +99,11 @@ def __init__(self,
                 section_name = 'llama'
 
             if len(section_name) > 0:
-                self.gpu_count = parser.getint(section_name,
-                                               'tensor_para_size')
+                tp_cfg = parser.getint(section_name, 'tensor_para_size')
                 self.session_len = parser.getint(section_name, 'session_len')
+                if tp_cfg != 1 and tp_cfg != tp:
+                    print(f'[INFO] found tp={tp_cfg} in config.ini.')
```
Comment (tpoisonooo) [inline lmdeploy/turbomind/turbomind.py:105]: 换 loguru 或者 logger

**Review by tpoisonooo — APPROVED** [review]
文档需额外声明量化不能动态 tp

File: lmdeploy/turbomind/turbomind.py (hunk 1 above)
Comment (grimoire) [reply lmdeploy/turbomind/turbomind.py]: done

File: lmdeploy/turbomind/turbomind.py (hunk 2 above)
Comment (grimoire) [reply lmdeploy/turbomind/turbomind.py]: done

File: README.md [hunk 3]
```diff
@@ -88,9 +88,11 @@ docker run --gpus all --rm -v $(pwd)/workspace:/workspace -it openmmlab/lmdeploy
...
 ```
 
-```{note}
-When inferring with FP16 precision, the InternLM-7B model requires at least 15.7G of GPU memory overhead on TurboMind. It is recommended to use NVIDIA cards such as 3090, V100, A100, etc.
-```
+> **Note**
+> When inferring with FP16 precision, the InternLM-7B model requires at least 15.7G of GPU memory overhead on TurboMind. It is recommended to use NVIDIA cards such as 3090, V100, A100, etc.
+
+> **Note**
+> Tensor parallel can be used to perform inference on multiple GPUs. Add `--tp=<num_gpu>` on `deploy` to enable static TP or add `--tp=<num_gpu>` on `chat` to enable runtime TP.
```
Comment (lvhan028) [inline README.md:95]: We don't need to run `turbomind.deploy`, do we?

File: lmdeploy/turbomind/utils.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import logging
+from typing import Optional
+
+logger_initialized = {}
+
+
+def get_logger(name: str,
```
Comment (lvhan028) [inline lmdeploy/turbomind/utils.py:8]: May remove logger in `serve/turbomind/utils.py`

File: src/turbomind/models/llama/LlamaDecoderLayerWeight.cc [hunk 5]
```diff
@@ -99,25 +100,135 @@ void mallocWeights(LlamaDenseWeight<T>& weights, bool bias)
...
 
 template<typename T>
-void loadWeights(LlamaDenseWeight<T>& w, std::string prefix, int rank, FtCudaDataType model_file_type)
+void loadWeights(LlamaDenseWeight<T>& w,
+                 std::string          prefix,
+                 int                  rank,
+                 FtCudaDataType       model_file_type,
+                 size_t               tensor_para_size,
+                 int                  slice_dim   = 0,
+                 std::vector<size_t>  slice_shape = {})
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaDecoderLayerWeight.cc:109]: It seems `slice_shape` isn't used when `loadWeights` is called. Is there any concern about this argument?

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**